connect/runtime/src/test/java/org/apache/kafka/connect/integration/ConnectorHandle.java

```diff
@@ -189,6 +192,83 @@ public void awaitCommits(int timeout) throws InterruptedException {
         }
     }
 
+    /**
+     * Record that this connector has been started. This should be called by the connector under
+     * test.
+     *
+     * @see #expectedRestarts(int)
+     * @see #expectedStarts(int)
+     */
+    public void recordConnectorStart() {
+        startAndStopCounter.recordStart();
+    }
+
+    /**
+     * Record that this connector has been stopped. This should be called by the connector under
+     * test.
+     *
+     * @see #expectedRestarts(int)
+     * @see #expectedStarts(int)
+     */
+    public void recordConnectorStop() {
+        startAndStopCounter.recordStop();
+    }
+
+    /**
+     * Obtain a {@link RestartLatch} that can be used to wait until the connector using this handle
+     * and all tasks using {@link TaskHandle} have completed the expected number of restarts,
+     * starting the counts at the time this method is called.
+     *
+     * <p>A test can call this method, specifying the number of times the connector and tasks
+     * will each be stopped and started from that point (typically {@code expectedRestarts(1)}).
+     * The test should then change the connector or otherwise cause the connector to restart one or
+     * more times, and then can call {@link RestartLatch#await(long, TimeUnit)} to wait up to a
+     * specified duration for the connector and all tasks to be restarted the specified number of
+     * times.
+     *
+     * <p>This method tracks the number of times the connector and tasks are stopped
+     * <em>and</em> started. See also {@link #expectedStarts(int)}.
+     *
+     * @param expectedRestarts the expected number of restarts
+     * @return the latch; never null
+     * @see #expectedStarts(int)
+     */
+    public RestartLatch expectedRestarts(int expectedRestarts) {
+        List<RestartLatch> taskLatches = taskHandles.values()
+                                                    .stream()
+                                                    .map(task -> task.expectedRestarts(expectedRestarts))
```

Comment:
by doing this, we are assuming that a connector restart implies a task restart. that might not be always true. adding workers might restart some tasks but not others (esp., with the new rebalancing algorithm). do you think we should allow users to be able to increment the expected number of restarts for individual tasks. if so, does this dependency still hold?

Reply:
Sounds interesting, cc @kkonstantine

Reply:
I'd assume this is meant to be used to assert reconfiguration. But I agree that it can be abused or be misleading. Also is it correct to assume that `expectedRestarts` means at least that many?

Reply:
Yes, I've updated the JavaDoc to denote this is a minimum number, and as mentioned above I've removed the `expectedRestarts` methods and now we have `expectedStarts` and `expectedStops` (if you really care about stops).